wai-extra/Network/Wai/Test.hs

```diff
@@ -214,11 +216,14 @@ assertString :: HasCallStack => String -> Session ()
 assertString s = unless (null s) $ assertFailure s
 
 assertFailure :: HasCallStack => String -> Session ()
-assertFailure msg = msg `deepseq` liftIO (throwIO (WaiTestFailure msg))
+assertFailure = liftIO . HUnit.assertFailure
 
-data WaiTestFailure = WaiTestFailure String
-    deriving (Show, Eq, Typeable)
-instance Exception WaiTestFailure
+{-# DEPRECATED WaiTestFailure "Use `HUnit.HUnitFailure` instead." #-}
```

Comment:
@snoyberg deprecation is optional.  I'm in favor of it for two reasons:

- The pattern synonym is not 100% behavior preserving (as mentioned before).
- Catching and then re-throwing `WaiTestFailure` will not preserve the source location.  Working with `HUnitFailure` directly does not have this limitation.

But not very opinionated on it, we can drop the deprecation if you prefer.